unittests/testanchorevents.py

```diff
@@ -0,0 +1,433 @@
+    "timestamp_accepted": "2022-10-14T09:30:52Z",
+    "timestamp_committed": "2022-10-14T09:30:52.920717507Z",
+    "timestamp_declared": "2019-11-28T14:44:19Z",
+    "from": "0x92a2923C46ff3904f8Fb0E80d1f0448AeE10BC9d",
+    "block_number": 0,
+    "transaction_id": "",
+    "transaction_index": 0,
+}
+
+NO_CONFIRMATION_EVENT = {
+    "identity": (
+        "assets/12349f7a-9d94-4ee0-835e-34567e882c9e/"
+        "events/97c7825a-a62e-459a-b091-4571cd7b467e"
+    ),
+    "asset_identity": "assets/12349f7a-9d94-4ee0-835e-34567e882c9e",
+    "behaviour": "RecordEvidence",
+    "operation": "Record",
+    "asset_attributes": {},
+    "event_attributes": {
+        "arc_description": "Safety conformance approved for version 1.6.",
+        "arc_evidence": "DVA Conformance Report attached",
+        "conformance_report": "blobs/e2a1d16c-03cd-45a1-8cd0-547189df1273",
+    },
+    "principal_accepted": {
+        "display_name": "stunt-idp@jitsuin.com",
+        "email": "stunt-idp@jitsuin.com",
+        "issuer": "stunt-idp@jitsuin.com",
+        "subject": "stunt-idp@jitsuin.com",
+    },
+    "principal_declared": {
+        "display_name": "",
+        "email": "phil.b@synsation.io",
+        "issuer": "idp.synsation.io/1234",
+        "subject": "phil.b",
+    },
+    "tenant_identity": "tenant/0dc854f5-5eab-4ad8-9d75-1ea9b24d0db4",
+    "timestamp_accepted": "2022-10-14T09:30:52Z",
+    "timestamp_committed": "2022-10-14T09:30:52.920717507Z",
+    "timestamp_declared": "2019-11-28T14:44:19Z",
+    "from": "0x92a2923C46ff3904f8Fb0E80d1f0448AeE10BC9d",
+    "block_number": 0,
+    "transaction_id": "",
+    "transaction_index": 0,
+}
+
+# these are deliberately in the wrong order sorted bu=y identity
+# so we can check the redact_sorted_events() function reorders them
+VALID_EVENTS_RESPONSE = {"events": VALID_EVENTS}
+VALID_EVENT0_RESPONSE = {"events": [VALID_EVENTS[0]]}
+VALID_EVENT1_RESPONSE = {"events": [VALID_EVENTS[1]]}
+
+VALID_EVENTS_EXPECTED_HASH = (
+    "61211c916cd113a1cf424ac729924de46aa6259919825dbdf8ec78c5c14665e2"
+)
+
+NO_EVENTS_EXPECTED_HASH = (
+    "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
+)
+
+PENDING_EVENTS_RESPONSE = {
+    "events": [
+        PENDING_EVENT,
+        VALID_EVENT,
+    ]
+}
+
+INCOMPLETE_EVENTS_RESPONSE = {
+    "events": [
+        INCOMPLETE_EVENT,
+        VALID_EVENT,
+    ]
+}
+
+NO_CONFIRMATION_EVENTS_RESPONSE = {
+    "events": [
+        NO_CONFIRMATION_EVENT,
+        VALID_EVENT,
+    ]
+}
+NO_EVENTS_RESPONSE = {
+    "events": [],
+}
+
+
+class TestHashEventsV1(TestCase):
+    """
+    Test anchor_events
+    """
+
+    maxDiff = None
+
+    def test_anchor_events_v1(self):
+        """
+        Test anchor_events
+        """
+        with mock.patch("requests.get") as mock_get:
+            mock_get.return_value = MockResponse(200, **VALID_EVENTS_RESPONSE)
+
+            simplehash = anchor_events(MIN_ACCEPTED, MAX_ACCEPTED, FQDN, AUTH_TOKEN)
```

Comment:
ah the nit that I was going to pick re MAX_ACCEPTED is that our anchor range is half open. so MAX_ACCEPTED is technically excluded. but that is only going to confuse one or two very observent maintainers

Reply:
I agree :) as does @eccles will leave in